### ISSUE TITLE
Wiremod usermessages to Net Library

### ISSUE THREAD

**Issue by Mantas-2155X** [wiremod/wire#946]
How about converting wiremod's usermessages to net library?
if you use EGP alot, there will be a crap ton of usermessages. Maybe change to net library?

**Comment by oldmud0** [comment]
What are the performance improvements?

**Comment by thegrb93** [comment]
http://facepunch.com/showthread.php?t=1304676

I've had enormous server/network lag from some egp games some people like to make. Net messages can help that, but something needs to be put in place to limit the rate of data sent or reliable channel overflows can occur.

**Comment by Nebual** [comment]
I believe EGP does actually use Net messages, https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_egp/lib/egplib/umsgsystem.lua , at least in most places, however it still sends 1 Net message per instruction.

You could definitely save some performance by switching EGP's use of packets to a more batch style, where it groups all requests in a single frame and then sends them as a single Net packet. Many parts of wire do this already.

**Comment by Divran** [comment]
Wiremod already uses net messages pretty much everywhere where it matters.

EGP already uses net messages, and it already tries to put as much data as it can in each message before sending it and starting a new one. When I made EGP, I tested it a lot to make sure the default transfer rates would not cause lag even if you spammed like shit. Someone else updated EGP to use net messages. Maybe that person did not properly make sure the default transfer rates didn't cause lag? I'm pretty sure they did though.

**Comment by thegrb93** [comment]
It was a long time ago I noticed it so could be different now.

**Comment by WGHydrogen** [comment]
EGP_HUD_Use:
EGP_ScrWH_Request:
HUDIndicatorFactor:
HUDIndicatorHideHUD:
HUDIndicatorRegister:
HUDIndicatorStyleFullCircle:
HUDIndicatorStylePercent:
HUDIndicatorUnRegister:
RcvEntityVarBeam_Angle:
RcvEntityVarBeam_Bool:
RcvEntityVarBeam_Entity:
RcvEntityVarBeam_Float:
RcvEntityVarBeam_Int:
RcvEntityVarBeam_String:
RcvEntityVarBeam_Vector:
UpdateEyePodState:
WireMapInterfaceEnt:
cpulib_debugdata_interrupt:
cpulib_debugdata_registers:
cpulib_debugdata_variables:
e2_remoteupload_request:
wire pod hud:
wire_clutch_links:
wire_gpu_memorymodel:
wire_gpu_monitorstate:
wire_gpulib_setent:
wire_memsync:
wire_spu_memorymodel:
wire_spu_soundsources:
wire_spu_soundstate:

Here are most of the wiremod's usermessages. Quite a few, huh?
How about converting these to net msg

**Comment by Divran** [comment]
`EGP_HUD_Use` and `EGP_ScrWH_Request` send less than 255 bytes of data, so converting them to net message is pointless. Same goes for `e2_remoteupload_request`. Most likely `wire pod hud` as well. I don't know about the others.

**Comment by WGHydrogen** [comment]
well, it would still be better to convert them for faster usage since net messages are faster than usermessages.

**Comment by Divran** [comment]
The speed difference doesn't matter in those cases that I listed. It would only be worth it if net messages are more efficient (by that I mean they put less strain on the server), but I've not seen any proof of that.

**Comment by WGHydrogen** [comment]
Well the usermessages spike like this: http://i.imgur.com/UXMxfaX.jpg

while using wiremod in our server.

**Comment by WGHydrogen** [comment]
The GPU ones look pretty big,

        umsg.Start("wire_gpu_memorymodel")
          umsg.Long(self:EntIndex())
          umsg.Long (self.RAMSize)
          umsg.Float(self.SerialNo)
          umsg.Short(self.ChipType)
        umsg.End()

  umsg.Start("wire_gpu_monitorstate")
    umsg.Long(self:EntIndex())
    umsg.Short(#self.Monitors)
    for idx=1,#self.Monitors do
      umsg.Long(self.Monitors[idx])
    end
  umsg.End()

**Comment by Divran** [comment]
I think net messages might show up on the net graph as usermessages, too. In any case, it's not possible for the ones I listed to cause such lag, since they are very rarely used. However I can't speak for the GPU/CPU ones since I don't know how those work.

Those might look big, but they're still obviously less than 255 bytes, since they only use a single usermessage.

**Comment by WGHydrogen** [comment]
No, net messages don't show up as usermessages in net_graph. But anyway, thank you for reviewing. If you ever think of converting them to net messages, please do so!

**Comment by TomyLobo** [comment]
wire_gpu_monitorstate looks like it would overflow the packet if you have like 60-ish GPUs

**Comment by TomyLobo** [comment]
https://github.com/TomyLobo/picker/blob/master/lua/autorun/client/umsg_stats.lua i made this a while back
if it still works, you might be able to use it to diagnose which umsg causes those spikes

**Comment by WGHydrogen** [comment]
Thank you! I will check it out.

**Comment by WGHydrogen** [comment]
Okay, so the issue is these usermessages:
RcvEntityVarBeam_Angle
RcvEntityVarBeam_Bool
RcvEntityVarBeam_Entity
RcvEntityVarBeam_Float
RcvEntityVarBeam_Int
RcvEntityVarBeam_String
RcvEntityVarBeam_Vector

They are the ones that are spiking. Any ideas?

**Comment by Divran** [comment]
ugh, that file. All dependencies on that file needs to be erased and the file deleted. I believe it's only used in a few places. I'd find out, but the github search function is complete and utter crap. Let me try google's `site:github.com` instead, that might give better results... Nope. I can't find where those functions are used without downloading wiremod (don't have it on this computer) and using notepad++'s find in files or something. Can you do it?

**Comment by WGHydrogen** [comment]
So you're saying it's safe to remove the file from my server?

**Comment by Divran** [comment]
No. It's obviously used somewhere or you wouldn't get the spikes. I'm saying we, the wire team, should phase it out and get rid of it because it's an ancient relic

**Comment by WGHydrogen** [comment]
Okay, I'm looking through all wiremod's lua files and I found this..

function ENT:Setup(gmode, draw_background)
	self.gmode = gmode
	self.outputMode = gmode
	self.draw_background = draw_background
	self:SetNetworkedBeamBool("draw_background", draw_background, true)
end

in wire/lua/entities/gmod_wire_graphics_tablet.lua

Gonna keep looking.

**Comment by WGHydrogen** [comment]
http://i.imgur.com/ggITnRp.jpg

**Comment by TomyLobo** [comment]
@Divran don't forget printing it out and burning the paper

**Comment by AbigailBuccaneer** [comment]
After #956 is pulled, the only thing that Wiremod still uses beam vars for is the wire paths. They don't change very often, but there's a lot of data in them. Replacing wire paths with the net library won't be a trivial task.

In particular, one thing the beam library does nicely is that it sends all entity data slowly when a client connects. Having a centralised system to do that does make a lot of sense, and if I were to switch everything to use NWVars then new player join times might be unreasonably long.

Still, I suspect a small custom system (using the net library rather than usermessages) just for wire paths would be quite a bit more efficient.

**Comment by Divran** [comment]
>In particular, one thing the beam library does nicely is that it sends all entity data slowly when a client connects.

Thing is, net messages are so huge that the odds you'd ever need more than a single one to send all the data is pretty small.

**Comment by WGHydrogen** [comment]
okay i took the files @AbigailBuccaneer edited and added them to my server. It seems it's better now. But some stuff still do it.

**Comment by TomyLobo** [comment]
can't you just move beam to net with the messages they way they are?

**Comment by Divran** [comment]
@TomyLobo that would be an enormous waste and we would gain absolutely nothing

**Comment by TomyLobo** [comment]
no I meant without touching those well-tested numbers.
and we would gain a more responsive reliable channel

**Comment by AbigailBuccaneer** [comment]
@TomyLobo that is true, it would be a partial solution. But the wire path messages are highly-structured data that's being sent really inefficiently as is, and as it's the only thing remaining using the beam library I think it makes sense to just replace it with something smaller and more specialised.

**Comment by Divran** [comment]
@TomyLobo If we don't change the values, we'd get the same amount of messages sent, just using a different library to send them. So the question is: what causes more lag? One large single message, or hundreds of tiny ones? I believe hundreds of tiny ones causes more lag, which is why I think it's a bad idea to simply change umsg to net and change nothing else.

In fact, it might be better to queue all the data and send it every 5 seconds to everyone. That way we can get even fewer (and larger) messages.
Except for the user creating the wires. They of course want to be able to see the wires immediately, so why not take advantage of the fact that adv wire is client side and simply create the wires immediately for that person? Legacy wire tool users would then need to wait up to 5 seconds for the next sync, in that case.

**Comment by oldmud0** [comment]
For some reason I always had the impression that Source used TCP, but it actually uses UDP.

5 seconds seems too long for a client sync... When we fiddle with packet sizes this is latency vs. throughput.

In this case, a mixed system would work. If we expect only a few messages then it would be fine to deliver them immediately so that latency does not become an issue. But say if we're pasting a dupe, then it would be better to buffer all the messages and send them as one packet.

How about a buffer that is flushed every 500ms? That would be a good tradeoff.

**Comment by TomyLobo** [comment]
@AbigailBuccaneer wrote:
21:48 - abigaildanger: https://github.com/wiremod/wire/issues/946 in response to this: i've written a nice lil replacement for wire paths, should have a much much lower bandwidth use (both in size of data and number of netmessages)
21:48 - abigaildanger: but i wrote it while really fuckin high
21:48 - abigaildanger: and so i'm not committing it even to a branch before i've read it while sober
21:48 - abigaildanger: and i mean i'd do that now but i'm also incredibly high rn

**Comment by Divran** [comment]
yep I saw that too

**Comment by WGHydrogen** [comment]
http://imgur.com/27Lqzwe
right.. so it seems the wire_memsync thing is literally spamming my server..
Any idea on this? I looked through the code, converted all the usermessages from wiremod except wire_memsync and beamvars to net messages, but this one is weird, it has all the data compress things and i really have no idea how to convert that.

**Comment by TomyLobo** [comment]
@WGHydrogen you're aware of the umsg_stats command?
It's supposed to show averages from the past, sorted by amount. You should try that.

**Comment by WGHydrogen** [comment]
@TomyLobo the umsg_stats doesn't really work..

**Comment by WGHydrogen** [comment]
Anyway, I have just spent an hour working to convert the beam_netvars to net messages.. and it's a success! Now its all net messages no more usermessages. The last thing is wire_memsync.

**Comment by Divran** [comment]
but didn't abigail already do that

**Comment by WGHydrogen** [comment]
no. I have converted the whole thing. Not only the few things he did.

**Comment by WGHydrogen** [comment]
http://imgur.com/YaUcWgP

**Comment by Divran** [comment]
ah nevermind, abigail actually REMOVED the beam var library from wiremod completely, which is even better.

**Comment by oldmud0** [comment]
Could changing everything to net messages make it more likely to cause a `buffer overflow in net message` error?

**Comment by WGHydrogen** [comment]
I haven't seen a buffer overflow in my server yet, wiremod has been actively used and its fine. I also made it so it sends the net messages instantly, instead of delayed like previously it was with umsg (beamvars)

**Comment by Divran** [comment]
it's still better to simply remove the beam vars library entirely

**Comment by WGHydrogen** [comment]
okay, if we did that, we would have to use networkvars right? But aren't NWVars and NetworkVars usermessages? Would it be faster to use net messages instead?

**Comment by TomyLobo** [comment]
Turning the beam vars lib into a shim around nwvars would give us the benefit of being able to easily switch to a new system if it makes sense

**Comment by Divran** [comment]
Pretty sure NWVars and NetworkVars are not usermessages. TomyLobo, that
would be true if wiremod used beam vars at all, but it only uses it in like
2 places.

**Comment by WGHydrogen** [comment]
well.. If i use any of the wiremod tools like: Screens, text screens, E2, GPU, CPU, EGP.. any wiremod tool, it sends some beamvars.

**Comment by Divran** [comment]
I'm pretty sure that's impossible. The beam library is only used in 2 places: the screen entity and graphics tablet. I would confirm this, just to be sure, but the GitHub search in repository feature sucks dick and couldn't find the ocean if it was standing in it. I don't have wiremod on this computer and I can't download it to do a find-in-files with a text editor, sooo I can't check it.

**Comment by WGHydrogen** [comment]
Screen: http://imgur.com/8rYYJAS
E2: http://imgur.com/guxe8xa
GPU: http://imgur.com/SO0NXpe

**Comment by Divran** [comment]
I just checked E2: https://github.com/wiremod/wire/tree/master/lua/entities/gmod_wire_expression2 it's not used anywhere in these three files, and I already know for a fact that it isn't used in any of the other files in E2.

**Comment by Divran** [comment]
Alright finally. What you're seeing is this https://github.com/wiremod/wire/blob/master/lua/wire/server/wirelib.lua#L711-L716 which is used to sync the wired inputs/outputs for rendering to the client. I believe @AbigailBuccaneer was going to (or already has) write a replacement for this along with removing the BeamVars library

**Comment by AbigailBuccaneer** [comment]
I suspect that the screen would've been the biggest bandwidth eater. However, I'm still going to replace the wire paths with something less awful.

There's a very WIP replacement for it at 20de2ce309646dbb2f87842eac5bee88243c5919.

**Comment by Garbius** [comment]
[Clutch stool] (https://github.com/wiremod/wire/blob/master/lua/wire/stools/clutch.lua) still uses umsgs to send link information. It's not done very often, but I guess it deserves to be mentioned.

**Comment by AbigailBuccaneer** [comment]
Thanks @Garbius. That should probably be using @Nebual's link stuff instead, which [uses net messages in a standardised way](https://github.com/wiremod/wire/blob/082a58dd176e7e51c8a7a2f0a079e48917ffb132/lua/wire/wireshared.lua#L707). I'll make a separate issue for it.
